In-app analytics module and initial instrumentation

adamlofting opened the pull request:
WIP - do not merge.

This could (can eventually) be abstracted out to a more generic module, but I wanted to get something working first.

To see this data arriving in GA, you need to look at the `Mobile App` view under the `Webmaker App` property.

Todo Notes:
- [x] Use custom dimensions to note which 'hits' happened while offline and were processed later
- [x] consider setting delays of ? mins before retrying if someone is offline
- [x] fix current issue where two events recorded simultaneously don't play nice getting into localForage
- [ ] Setup conversion goals in GA to use publishing events we're firing

Future improvements
- We could do with some more friendly screen names through the app if we want to surface these in GA

adamlofting commented:
cc/ @thisandagain

adamlofting commented:
New features:
* If a user is offline, it holds off trying to reach GA for 5 minutes to limit unnecessary pings
* I fixed a race conditions bug trying to append multiple logging items (hits) to the localForage queue while offline

adamlofting commented:
also cc @davidascher @k88hudson

thisandagain commented:
Nice start. Thanks for staging this PR @adamlofting

davidascher commented:
Are there 'user-online'/'user-offline' hooks we can use to drive ping
schedule?

thisandagain commented:
The hooks within Cordova `org.apache.cordova.network-information` might be a good candidate:
http://docs.phonegap.com/en/4.0.0/cordova_events_events.md.html

thisandagain commented:
Hey @adamlofting any updates?

thisandagain commented:
@adamlofting Heads-up:
We added a new module at `./lib/network` that should help you detect online / offline states.

adamlofting commented:
@thisandagain thanks for that network lib, that's very useful.

I've made progress on this today:
* Cleaned up the logic for offline queuing, and fixed bugs
* Instrumented a bunch of tracking points across the app

I've made use of the `.lib/network` here:
https://github.com/mozilla/webmaker-app/pull/904/files#diff-fbd9fde296995083e4d606b8d6e31a50R101

But I haven't been able to get debugging working for Android today. (adb isn't recognizing my galaxy S2, and I couldn't get Weinre to recognize the emulator). 

So, if you have a chance to test this against an Android Cordova build, that would be great.

thisandagain commented:
Can do. Do you mind resolving the lint errors so tests pass?

adamlofting commented:
Lint errors should be resolved now. :smile:  
I wasn't sure how to get around the 'used before it was declared' ordering of two functions that call each other recursively but went with this which seemed reasonable: http://www.tumuski.com/2009/01/jslint/

thisandagain commented on components/inlineEditor/index.js:
```diff
@@ -29,6 +31,8 @@ module.exports = {
             this.$dispatch('inlineEditorStarted', {index: this.$index});
             this.$el.classList.add('active');
             this.centerEditor();
+            analytics.event({category: 'Inline Editor',
```
The max column width has been extended to 128, feel free to collapse.

thisandagain commented on lib/analytics.js:
```diff
@@ -0,0 +1,405 @@
+var packageJSON = clone(config.package);
+delete config.package;
+
+var PRE_QUEUE = [];
```
We use `camelCase` for all variables.

thisandagain commented on lib/analytics.js:
```diff
@@ -0,0 +1,405 @@
+    }, function (err, resp, body) {
+        if (resp.statusCode === 200) {
+            // we're got as far as GA (and must be online)
+            _log('sent hit to GA', connectionText);
```
This is going to be noisy when debugging.

adamlofting replied:
Good call. This is now off by default, except for warnings. It can be enabled with `ANALYTICS_CONSOLE_LOGGING`

thisandagain commented on lib/analytics.js:
```diff
@@ -0,0 +1,405 @@
+ * Functions for queuing offline hits, and processing those
+ * queues when connectivity is available
+ *
+ * A note on queues: The offline queue is stored using the localForage
```
Ooof. Hopefully the worker / IndexedDB issues will be resolved in Firefox / Firefox OS soon as to make this simpler.

thisandagain commented on lib/analytics.js:
```diff
@@ -0,0 +1,405 @@
+function logQueueLength () {
+    localForage.getItem('gaQueue', function (err, queue) {
+        queue = queue || [];
+        _log('Queue length: ' + queue.length);
```
This also gets a bit noisy in the logs.

thisandagain commented on views/make/index.js:
```diff
@@ -72,6 +73,9 @@ module.exports = view.extend({
             app.update({
                 name: newVal
             });
+            analytics.event({ category: 'App Name & Icon',
```
When you pull from upstream you'll notice that this no longer autosaves.

thisandagain commented:
Really exceptional work @adamlofting . This looks incredible. I left a few minor notes in-line. Major things that I'd like to address prior to 1.0 (not a blocker for merging this):
- Modularity: Split the additions to `util` and the `analytics` module into a distinct `npm` module
- Test Coverage: Setup unit and browser test coverage
- Documentation: Document the public API for the `analytics` module
- Performance: The in-memory `PRE_QUEUE` is an interesting work-around to running IndexedDB within a web worker, but I think we should let `localforage` abstract that for us as to keep the codebase minimal (https://github.com/mozilla/localForage/issues/144). Thoughts?

adamlofting commented:
:hatched_chick:

Inline notes have all been addressed now, and I've rebased and squashed the commits.

I'd love to see this merged so we can start seeing test data in GA from dev usage. And I'll start work on documentation now so that devs can continue to instrument things as we build out features.

And :+1: to all suggested future improvements to this.

k88hudson commented:
This is super awesome :star2: :star2: :star2: 

Once this is ready to go would you mind adding some brief documentation about how to add event tracking to https://github.com/mozilla/webmaker-app/tree/master/_docs? Just to create a new analytics folder and add a README.md, then add an item in the table of contents here https://github.com/mozilla/webmaker-app/blob/master/SUMMARY.md
